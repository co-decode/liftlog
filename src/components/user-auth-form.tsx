"use client";

import * as React from "react";
import { zodResolver } from "@hookform/resolvers/zod";
import { signIn } from "next-auth/react";
import { useForm } from "react-hook-form";
import * as z from "zod";

import { cn } from "@/lib/utils";
import { buttonVariants } from "@/components/ui/button";
import { Input } from "@/components/ui/input";
import { Label } from "@/components/ui/label";
import { toast } from "@/components/ui/use-toast";
import { Icons } from "@/components/icons";
import { useRouter } from "next/router";
import { UserRegForm } from "./user-reg-form";

interface UserAuthFormProps extends React.HTMLAttributes<HTMLDivElement> {
  register?: boolean
}

const userAuthCredSchema = z.object({
  identifier: z.string(),
  password: z.string(),
});
type FormDataCred = z.infer<typeof userAuthCredSchema>;

export function UserAuthForm({ register, className, ...props }: UserAuthFormProps) {
  const router = useRouter();
  const credSchema = useForm<FormDataCred>({
    resolver: zodResolver(userAuthCredSchema),
  });
  const [isLoading, setIsLoading] = React.useState<boolean>(false);
  const [isGitHubLoading, setIsGitHubLoading] = React.useState<boolean>(false);
  const [isNetlifyLoading, setIsNetlifyLoading] =
    React.useState<boolean>(false);

  if (register) return <UserRegForm
    isLoading={isLoading}
    setIsLoading={setIsLoading}
    isGitHubLoading={isGitHubLoading}
    setIsGitHubLoading={setIsGitHubLoading}
    isNetlifyLoading={isNetlifyLoading}
    setIsNetlifyLoading={setIsNetlifyLoading}
  />

  async function onSubmitCreds(data: FormDataCred) {
    setIsLoading(true);

    const signInResult = await signIn("credentials", {
      identifier: data.identifier.toLowerCase(),
      password: data.password,
      redirect: false,
      callbackUrl: "/dashboard",
    });

    if (!signInResult?.ok) {
      setIsLoading(false);
      return toast({
        title: "Something went wrong.",
        description: "Your sign in request failed. Please try again.",
        variant: "destructive",
      });
    }

    return router.push("/dashboard")
  }

  return (
    <div className={cn("grid gap-6", className)} {...props}>
      <form onSubmit={credSchema.handleSubmit(onSubmitCreds)}>
        <div className="grid gap-2">
          <div className="grid gap-2">
            <Label className="sr-only" htmlFor="identifier">
              Email
            </Label>
            <Input
              id="identifier"
              placeholder="Email"
              type="text"
              autoCapitalize="none"
              autoCorrect="off"
              disabled={isLoading || isGitHubLoading || isNetlifyLoading}
              {...credSchema.register("identifier")}
            />
            {credSchema.formState.errors?.identifier && (
              <p className="px-1 text-xs text-red-600">
                {`${credSchema.formState.errors.identifier?.message}`}
              </p>
            )}
            <Label className="sr-only" htmlFor="email">
              Password
            </Label>
            <Input
              id="password"
              placeholder="Password"
              type="password"
              autoCapitalize="none"
              autoCorrect="off"
              disabled={isLoading || isGitHubLoading || isNetlifyLoading}
              {...credSchema.register("password")}
            />
            {credSchema.formState.errors?.password && (
              <p className="px-1 text-xs text-red-600">
                {`${credSchema.formState.errors.password.message}`}
              </p>
            )}
          </div>
          <button className={cn(buttonVariants())} disabled={isLoading}>
            {isLoading && (
              <Icons.spinner className="mr-2 h-4 w-4 animate-spin" />
            )}
            Sign In with Credentials
          </button>
      <button className={cn(buttonVariants({variant: "destructive"}))} disabled={isLoading}>
        {isLoading && (
          <Icons.spinner className="mr-2 h-4 w-4 animate-spin" />
        )}
        Sign In with a Trial Account
      </button>
        </div>
      </form>
      <div className="relative">
        <div className="absolute inset-0 flex items-center">
          <span className="w-full border-t" />
        </div>
        <div className="relative flex justify-center text-xs uppercase">
          <span className="bg-background px-2 text-muted-foreground">
            Or continue with Email
          </span>
        </div>
      </div>
      <UserRegForm
        isLoading={isLoading}
        setIsLoading={setIsLoading}
        isGitHubLoading={isGitHubLoading}
        setIsGitHubLoading={setIsGitHubLoading}
        isNetlifyLoading={isNetlifyLoading}
        setIsNetlifyLoading={setIsNetlifyLoading}
      />
    </div>
  );
}
/*
<form onSubmit={emailSchema.handleSubmit(onSubmitEmail)}>
<div className="grid gap-2">
<div className="grid gap-2">
<Label className="sr-only" htmlFor="email">
Email
</Label>
<Input
id="email"
placeholder="name@example.com"
type="email"
autoCapitalize="none"
autoComplete="email"
autoCorrect="off"
disabled={isLoading || isGitHubLoading || isNetlifyLoading}
{...emailSchema.register("email")}
/>
{emailSchema.formState.errors?.email && (
    <p className="px-1 text-xs text-red-600">
    {`${emailSchema.formState.errors.email.message}`}
    </p>
    )}
</div>
<button className={cn(buttonVariants())} disabled={isLoading}>
{isLoading && (
    <Icons.spinner className="mr-2 h-4 w-4 animate-spin" />
    )}
Sign In with Email
</button>
</div>
</form>
<div className="relative">
<div className="absolute inset-0 flex items-center">
<span className="w-full border-t" />
</div>
<div className="relative flex justify-center text-xs uppercase">
<span className="bg-background px-2 text-muted-foreground">
Or continue with
</span>
</div>
</div>
<div className="flex flex-col gap-2">
<button
type="button"
className={cn(buttonVariants({ variant: "outline" }))}
onClick={() => {
  setIsGitHubLoading(true);
  signIn("github", { redirect: false });
}}
disabled={isLoading || isGitHubLoading || isNetlifyLoading}
>
{isGitHubLoading ? (
    <Icons.spinner className="mr-2 h-4 w-4 animate-spin" />
    ) : (
      <Icons.gitHub className="mr-2 h-4 w-4" />
      )}{" "}
    Github
    </button>
    <button
    type="button"
    className={cn(buttonVariants({ variant: "outline" }), "p-0")}
    onClick={() => {
      setIsNetlifyLoading(true);
      signIn("netlify", { redirect: false });
    }}
disabled={isLoading || isGitHubLoading || isNetlifyLoading}
>
{isNetlifyLoading ? (
    <Icons.spinner className="mr-2 h-4 w-4 animate-spin" />
    ) : (
      <Icons.netlify className="text-white mr-2 h-4 w-4" />
      )}{" "}
    Netlify
    </button>
    </div>
    */
/*
  async function onSubmitEmail(data: FormDataEmail) {
    setIsLoading(true);

    const signInResult = await signIn("email", {
      email: data.email.toLowerCase(),
      redirect: false,
      callbackUrl: searchParams?.get("from") || "/dashboard",
    });

    setIsLoading(false);

    if (!signInResult?.ok) {
      return toast({
        title: "Something went wrong.",
        description: "Your sign in request failed. Please try again.",
        variant: "destructive",
      });
    }

    return toast({
      title: "Check your email",
      description: "We sent you a login link. Be sure to check your spam too.",
    });
  }
  */
/*
  const emailSchema = useForm<FormDataEmail>({
    resolver: zodResolver(userAuthEmailSchema),
  });
  */
//type FormDataEmail = z.infer<typeof userAuthEmailSchema>;
/*
const userAuthEmailSchema = z.object({
  email: z.string().email(),
});
*/
