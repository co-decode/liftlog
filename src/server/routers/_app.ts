import { z } from 'zod';
import { Prisma } from '@prisma/client';
import { router, procedure } from '../trpc';
import { prisma } from '../prisma';

const allUser = Prisma.validator<Prisma.UserSelect>()({
    name: true,
    id: true,
    email: false,
    emailVerified: false,
});

export const appRouter = router ({
    findAll: procedure
    .query(async () => {
            return await prisma.user.findMany({
                select: allUser,
            });
    }),
    insertOne: procedure
    .input(
        z.object({
            username: z.string(),
            password: z.string(),
        }),
    )
    .mutation(async ({input}) => {
        return await prisma.user.create({
            data: { 
                name: input.username,
            },
        });
    }),
    updateOne: procedure
    .input (
        z.object({
            id: z.number(),
            username: z.string(),
            password: z.string(),
        }),
    )
    .mutation(async ({input}) => {
        const { id, ...rest } = input;
        return await prisma.user.update({
            where: { id },
            data: { ...rest },
        });
    }),
    deleteAll: procedure
    .input (
        z.object({
            ids: z.number().array(),
        }),
    )
    .mutation(async ({input}) => {
        const { ids } = input;

        return await prisma.user.deleteMany({
            where: {
                id: { in: ids },
            },
        });
    }),
})

// export type definition of API
export type AppRouter = typeof appRouter;
