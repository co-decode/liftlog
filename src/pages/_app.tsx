import "@/styles/globals.css";
import "@/styles/fonts.module.css";
import type { AppProps } from "next/app";
import type { Session } from "next-auth";
import Head from "next/head";
import { ThemeProvider } from "@/components/theme-provider";
import { trpc } from "../utils/trpc";
import { SessionProvider } from "next-auth/react";
import ContextProvider from "@/components/auth-and-context"
import { Inter } from "next/font/google";

const fontSans = Inter({
  subsets: ["latin"],
  variable: "--font-sans",
});

const App = ({
  Component,
  pageProps: { session, ...pageProps },
}: AppProps<{ session: Session }>) => {
  return (
    <>
      <Head>
        <title>LiftLog</title>
        <meta
          name="viewport"
          content="minimum-scale=1, initial-scale=1, width=device-width"
        />
        <link rel="shortcut icon" href="/favicon.svg"/>
      </Head>
      <style jsx global>
        {`
          html {
            font-family: ${fontSans.style.fontFamily};
          }
        `}
      </style>

      <SessionProvider session={session}>
        <ThemeProvider attribute="class" defaultTheme="system" enableSystem>
        <ContextProvider>
          <Component {...pageProps} />
        </ContextProvider>
        </ThemeProvider>
      </SessionProvider>
    </>
  );
};

export default trpc.withTRPC(App);
