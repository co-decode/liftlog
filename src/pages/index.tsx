import { Button, buttonVariants } from "@/components/ui/button"
import { MainNav } from "@/components/main-nav";
import { cn } from "@/lib/utils";
import localFont from "next/font/local"
import { useSession } from "next-auth/react";
import Link from "next/link";
import { useRouter } from "next/router";
import { useEffect } from "react";
import { Icons } from "@/components/icons";

const fontHeading = localFont({
  src: "../../public/fonts/CalSans-SemiBold.woff2",
  variable: "--font-heading",
})

export default function IndexPage() {
  const { data: session, status } = useSession();

  const router = useRouter();
  useEffect(() => {
    if (session) {
      router.push('/dashboard');
    }
  }, [session, router]);

  return (
    <>
      <header className="container z-40 bg-background">
        <div className="flex h-20 items-center justify-between py-6">
          <MainNav />
          <nav>
            <Link
              aria-label="disableable"
              href="/login"
              className={cn(
                buttonVariants({ variant: "secondary", size: "sm" }),
                status === "loading" && "pointer-events-none",
                "px-4"
              )}
            >
              Login
            </Link>
          </nav>
        </div>
      </header>
      <section className="grid justify-items-center gap-2 sm:gap-3 lg:gap-4 py-5">
        <h1 className={cn("scroll-m-20 text-5xl font-heading md:text-6xl lg:text-7xl", fontHeading.className)}>
          LiftLog
        </h1>
        <p className="max-w-[42rem] leading-normal text-muted-foreground sm:text-xl sm:leading-8 pb-3">
          Log your Lifts with LiftLog.
        </p>
        <div className="space-x-4">
          <Link
            aria-label="disableable"
            href="/login"
            className={cn(buttonVariants({ size: "lg" }), status === "loading" && "pointer-events-none")}>
            Get Started
          </Link>
          <Link
            aria-label="disableable"
            href={"https://github.com/co-decode/liftlog"}
            target="_blank"
            rel="noreferrer"
            className={cn(buttonVariants({ variant: "outline", size: "lg" }), status === "loading" && "pointer-events-none")}
          >
            GitHub
          </Link>
        </div>
        <Link href="/login"
          aria-label="disableable"
          className={cn(buttonVariants({ variant: "destructive", size: "lg" }),
            status === "loading" && "pointer-events-none",
            "w-[271px]")}
        >
          Trial with a Test Account
        </Link>
      </section>
      <section id="about" className="container py-8 md:py-12 lg:py-24">
        <div className="mx-auto flex max-w-[58rem] flex-col items-center justify-center gap-4 text-center">
          <h2 className={cn("text-3xl leading-[1.1] sm:text-3xl md:text-6xl", fontHeading.className)}>
            Check it out
          </h2>
          <p className="max-w-[85%] leading-normal text-muted-foreground sm:text-lg sm:leading-7">
            View a demonstration of LiftLog&apos;s features.
          </p>
          <Link
            aria-label="disableable"
            href="https://cody-ross.net/liftlog"
            className={cn(status === "loading" && "pointer-events-none")}
            target="_blank"
          >
            <Button
              disabled={status === "loading"}
              role="link"
            >
              Demonstration
              <Icons.arrowRight className="h-4 w-4 ml-2" />
            </Button>
          </Link>
        </div>
      </section >
      <section
        id="features"
        className="container space-y-6 bg-slate-50 py-6 dark:bg-transparent md:py-12 lg:py-24"
      >
        <div className="mx-auto flex max-w-[58rem] flex-col items-center space-y-4 text-center">
          <h2 className={cn("font-heading text-3xl leading-[1.1] sm:text-3xl md:text-6xl", fontHeading.className)}>
            Features
          </h2>
          <p className="max-w-[85%] leading-normal text-muted-foreground sm:text-lg sm:leading-7">
            This project is a free and open-source weightlifting application.
            Create workout programs and use them to perform and record workouts.
          </p>
        </div>
        <div className="mx-auto grid justify-center gap-4 sm:grid-cols-2 md:max-w-[64rem] md:grid-cols-3">
          <div className="relative overflow-hidden rounded-lg border bg-background p-2">
            <div className="flex h-[180px] flex-col justify-between rounded-md p-6">
              <Icons.dumbbell className="h-12 w-12" />
              <div className="space-y-2">
                <h3 className="font-bold">Training Programs</h3>
                <p className="text-sm text-muted-foreground">
                  Create a program to structure your training.
                </p>
              </div>
            </div>
          </div>
          <div className="relative overflow-hidden rounded-lg border bg-background p-2">
            <div className="flex h-[180px] flex-col justify-between rounded-md p-6">
              <Icons.timer className="h-12 w-12" />
              <div className="space-y-2">
                <h3 className="font-bold">Workout Aide</h3>
                <p className="text-sm">
                  Record your performance as you workout.
                </p>
              </div>
            </div>
          </div>
          <div className="relative overflow-hidden rounded-lg border bg-background p-2">
            <div className="flex h-[180px] flex-col justify-between rounded-md p-6">
              <Icons.calendar className="h-12 w-12" />
              <div className="space-y-2">
                <h3 className="font-bold">Schedule</h3>
                <p className="text-sm text-muted-foreground">
                  Track upcoming and previous workout sessions.
                </p>
              </div>
            </div>
          </div>
          <div className="relative overflow-hidden rounded-lg border bg-background p-2">
            <div className="flex h-[180px] flex-col justify-between rounded-md p-6">
              <Icons.graph className="h-12 w-12" />
              <div className="space-y-2">
                <h3 className="font-bold">Visualisation</h3>
                <p className="text-sm text-muted-foreground">
                  Analyse your training data and progress.
                </p>
              </div>
            </div>
          </div>
          <div className="relative overflow-hidden rounded-lg border bg-background p-2">
            <div className="flex h-[180px] flex-col justify-between rounded-md p-6">
              <Icons.list className="h-12 w-12" />
              <div className="space-y-2">
                <h3 className="font-bold">Log</h3>
                <p className="text-sm text-muted-foreground">
                  Filter and search through training sessions.
                </p>
              </div>
            </div>
          </div>
          <div className="relative overflow-hidden rounded-lg border bg-background p-2">
            <div className="flex h-[180px] flex-col justify-between rounded-md p-6">
              <svg
                viewBox="0 0 24 24"
                fill="none"
                stroke="currentColor"
                strokeWidth="2"
                className="h-12 w-12"
              >
                <path d="M12 22s8-4 8-10V5l-8-3-8 3v7c0 6 8 10 8 10z"></path>
              </svg>              <div className="space-y-2">
                <h3 className="font-bold">Authentication</h3>
                <p className="text-sm text-muted-foreground">
                  Multiple authentication flows for ease of use.
                </p>
              </div>
            </div>
          </div>
        </div>
      </section>
      {/*
    */}
    </>
  );
}
