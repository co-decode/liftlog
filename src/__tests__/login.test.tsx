import { getByRole, render, screen, cleanup, fireEvent, act, waitFor } from '@testing-library/react'
import type { Session } from "next-auth";
import Login from '../pages/login'
import { AppWrapper } from '../pages/_app'
import '@testing-library/jest-dom'
import { describe, test, expect } from '@jest/globals'
import { useRouter } from 'next/router';
import { useSession } from 'next-auth/react';
import { UserAuthForm } from '../components/user-auth-form'

const mockSession: Session = {
  user: {
    email: "mock@test.com",
    name: "Mocky Mouse",
    image: null
  },
  expires: '05 October 2024 14:48 UTC'

}

// @ts-ignore
jest.mock('next/router', () => ({
  // @ts-ignore
  useRouter: jest.fn()
}))

// mockSession is redefined because mocks are hoisted above import statements
// @ts-ignore
jest.mock("next-auth/react", () => {
  // @ts-ignore
  const originalModule = jest.requireActual('next-auth/react');
  const mockSession = {
    expires: new Date(Date.now() + 2 * 86400).toISOString(),
    user: { name: "admin", email: "admin@mock.com" }
  };
  return {
    ...originalModule,
    // @ts-ignore
    useSession: jest.fn(() => {
      return { data: mockSession, status: 'authenticated' }  // return type is [] in v3 but changed to {} in v4
    }),
    // @ts-ignore
    signIn: jest.fn(() => {
      return { ok: true }
    })
  };
});

describe('Index Page', () => {
  test('Authentication: Redirect if session already exists', () => {
    // setup a new mocking function for push method
    // @ts-ignore
    const pushMock = jest.fn()
    // mock a return value on useRouter
    // @ts-ignore
    useRouter.mockReturnValue({
      query: {},
      push: pushMock,
    })
    render(<AppWrapper session={mockSession} > <Login /></AppWrapper >)

    expect(pushMock).toHaveBeenCalledTimes(1)
    expect(pushMock).toHaveBeenCalledWith('/dashboard')
  })

  test('Match Snapshot', async () => {
    const { asFragment } = render(<AppWrapper session={mockSession} > <Login /></AppWrapper >)
    expect(asFragment()).toMatchSnapshot()
  })
})

