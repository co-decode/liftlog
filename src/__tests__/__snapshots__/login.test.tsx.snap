// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Index Page Match Snapshot 1`] = `
<DocumentFragment>
   
  <div
    class="min-h-screen grid grid-rows-[min-content_1fr]"
  >
    <header
      class="container z-40 bg-background"
    >
      <div
        class="flex h-16 items-center justify-between py-4"
      >
        <div
          class="flex gap-6 md:gap-10"
        >
          <a
            class="items-center space-x-2 flex"
            href="/dashboard"
          >
            <svg
              class="lucide lucide-dumbbell"
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="m6.5 6.5 11 11"
              />
              <path
                d="m21 21-1-1"
              />
              <path
                d="m3 3 1 1"
              />
              <path
                d="m18 22 4-4"
              />
              <path
                d="m2 6 4-4"
              />
              <path
                d="m3 10 7-7"
              />
              <path
                d="m14 21 7-7"
              />
            </svg>
            <span
              class="font-bold inline-block"
            >
              LiftLog
            </span>
          </a>
        </div>
        <nav>
          <a
            class="inline-flex items-center justify-center text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:opacity-50 disabled:pointer-events-none ring-offset-background bg-secondary text-secondary-foreground hover:bg-secondary/80 h-9 rounded-md px-3"
            href="/"
          >
            <svg
              class="mr-1 h-4 w-4"
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <polyline
                points="15 18 9 12 15 6"
              />
            </svg>
            Back
          </a>
        </nav>
      </div>
    </header>
    <div
      class="container flex w-screen flex-col items-center justify-center"
    >
      <div
        class="mx-auto flex w-full flex-col justify-center space-y-6 sm:w-[350px]"
      >
        <div
          class="flex flex-col space-y-2 text-center"
        >
          <svg
            class="mx-auto h-8 w-8"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="m6.5 6.5 11 11"
            />
            <path
              d="m21 21-1-1"
            />
            <path
              d="m3 3 1 1"
            />
            <path
              d="m18 22 4-4"
            />
            <path
              d="m2 6 4-4"
            />
            <path
              d="m3 10 7-7"
            />
            <path
              d="m14 21 7-7"
            />
          </svg>
          <h1
            class="text-2xl font-semibold tracking-tight"
          >
            Welcome back
          </h1>
          <p
            class="text-sm text-muted-foreground"
          >
            Enter your credentials to sign in
          </p>
        </div>
        <div
          class="grid gap-6"
        >
          <form>
            <div
              class="grid gap-2"
            >
              <div
                class="grid gap-2"
              >
                <label
                  class="text-sm font-medium leading-none peer-disabled:cursor-not-allowed peer-disabled:opacity-70 sr-only"
                  for="identifier"
                >
                  Email
                </label>
                <input
                  autocapitalize="none"
                  autocorrect="off"
                  class="flex h-10 w-full rounded-md border border-input bg-transparent px-3 py-2 text-sm ring-offset-background file:border-0 file:bg-transparent file:text-sm file:font-medium placeholder:text-muted-foreground focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:cursor-not-allowed disabled:opacity-50"
                  id="identifier"
                  name="identifier"
                  placeholder="Email"
                  type="text"
                />
                <label
                  class="text-sm font-medium leading-none peer-disabled:cursor-not-allowed peer-disabled:opacity-70 sr-only"
                  for="email"
                >
                  Password
                </label>
                <input
                  autocapitalize="none"
                  autocorrect="off"
                  class="flex h-10 w-full rounded-md border border-input bg-transparent px-3 py-2 text-sm ring-offset-background file:border-0 file:bg-transparent file:text-sm file:font-medium placeholder:text-muted-foreground focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:cursor-not-allowed disabled:opacity-50"
                  id="password"
                  name="password"
                  placeholder="Password"
                  type="password"
                />
              </div>
              <button
                aria-label="disableable"
                class="inline-flex items-center justify-center rounded-md text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:opacity-50 disabled:pointer-events-none ring-offset-background bg-primary text-primary-foreground hover:bg-primary/90 h-10 py-2 px-4"
              >
                Sign In with Credentials
              </button>
              <button
                aria-label="disableable"
                class="inline-flex items-center justify-center rounded-md text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:opacity-50 disabled:pointer-events-none ring-offset-background bg-destructive text-destructive-foreground hover:bg-destructive/90 h-10 py-2 px-4"
                role="button"
              >
                Sign In with a Trial Account
              </button>
            </div>
          </form>
          <div
            class="relative"
          >
            <div
              class="absolute inset-0 flex items-center"
            >
              <span
                class="w-full border-t"
              />
            </div>
            <div
              class="relative flex justify-center text-xs uppercase"
            >
              <span
                class="bg-background px-2 text-muted-foreground"
              >
                Or continue with Email
              </span>
            </div>
          </div>
          <div
            class="grid gap-6"
          >
            <form>
              <div
                class="grid gap-2"
              >
                <div
                  class="grid gap-2"
                >
                  <label
                    class="text-sm font-medium leading-none peer-disabled:cursor-not-allowed peer-disabled:opacity-70 sr-only"
                    for="email"
                  >
                    Email
                  </label>
                  <input
                    autocapitalize="none"
                    autocomplete="email"
                    autocorrect="off"
                    class="flex h-10 w-full rounded-md border border-input bg-transparent px-3 py-2 text-sm ring-offset-background file:border-0 file:bg-transparent file:text-sm file:font-medium placeholder:text-muted-foreground focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:cursor-not-allowed disabled:opacity-50"
                    id="email"
                    name="email"
                    placeholder="name@example.com"
                    type="email"
                  />
                </div>
                <button
                  class="inline-flex items-center justify-center rounded-md text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:opacity-50 disabled:pointer-events-none ring-offset-background bg-primary text-primary-foreground hover:bg-primary/90 h-10 py-2 px-4"
                >
                  Sign In with Email
                </button>
              </div>
            </form>
            <div
              class="relative"
            >
              <div
                class="absolute inset-0 flex items-center"
              >
                <span
                  class="w-full border-t"
                />
              </div>
              <div
                class="relative flex justify-center text-xs uppercase"
              >
                <span
                  class="bg-background px-2 text-muted-foreground"
                >
                  Or continue with
                </span>
              </div>
            </div>
            <div
              class="flex flex-col gap-2"
            >
              <button
                class="inline-flex items-center justify-center rounded-md text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:opacity-50 disabled:pointer-events-none ring-offset-background border border-input hover:bg-accent hover:text-accent-foreground h-10 py-2 px-4"
                type="button"
              >
                <svg
                  aria-hidden="true"
                  class="mr-2 h-4 w-4"
                  data-icon="github"
                  data-prefix="fab"
                  focusable="false"
                  role="img"
                  viewBox="0 0 496 512"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M165.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3 .3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6zm-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5 .3-6.2 2.3zm44.2-1.7c-2.9 .7-4.9 2.6-4.6 4.9 .3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9zM244.8 8C106.1 8 0 113.3 0 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C428.2 457.8 496 362.9 496 252 496 113.3 383.5 8 244.8 8zM97.2 352.9c-1.3 1-1 3.3 .7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1zm-10.8-8.1c-.7 1.3 .3 2.9 2.3 3.9 1.6 1 3.6 .7 4.3-.7 .7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3 .7zm32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3 .7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1zm-11.4-14.7c-1.6 1-1.6 3.6 0 5.9 1.6 2.3 4.3 3.3 5.6 2.3 1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2z"
                    fill="currentColor"
                  />
                </svg>
                 Github
              </button>
              <button
                class="inline-flex items-center justify-center rounded-md text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:opacity-50 disabled:pointer-events-none ring-offset-background border border-input hover:bg-accent hover:text-accent-foreground h-10 p-0"
                type="button"
              >
                <svg
                  aria-hidden="true"
                  class="mr-2 h-4 w-4"
                  data-icon="netlify"
                  fill="currentColor"
                  focusable="false"
                  height="800px"
                  role="img"
                  viewBox="0 0 32 32"
                  width="800px"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M22.579 11.36c0.151 0.077 0.291 0.181 0.405 0.307l3.131-1.396-2.921-2.896-0.657 3.943zM16.079 8.729c0.156 0.229 0.249 0.489 0.281 0.765l4.661 1.975c0.14-0.109 0.301-0.188 0.473-0.235l0.765-4.735-2.839-2.979-3.343 5.135zM31.989 16.047l-4.995-4.995-3.401 1.391 8.355 3.552c0 0 0.073 0.052 0.041 0.052zM31.156 16.86l-8.020-3.428c-0.224 0.297-0.563 0.495-0.933 0.541l-0.859 5.277c0.255 0.26 0.401 0.609 0.401 0.973l4.843 1.016 4.443-4.412v-0.083zM20.531 12.333l-4.353-1.855c-0.308 0.469-0.833 0.745-1.391 0.724h-0.265l-3.609 5.568 9.635-4.172v0.011c-0.005-0.073-0.005-0.145 0-0.224 0.005-0.016 0-0.036-0.016-0.052zM25.823 22.079l-4.26-0.881c-0.161 0.209-0.369 0.365-0.615 0.459l-1.016 6.297 5.917-5.865c0 0-0.011 0.032-0.027 0.032zM19.864 21.645c-0.411-0.167-0.713-0.52-0.823-0.948l-7.932-1.629-0.115 0.183 7.199 10.411 0.432-0.427 1.224-7.563c0 0 0.041 0.031 0.015 0.016zM11.276 18.109l7.839 1.615c0.229-0.401 0.64-0.661 1.099-0.697l0.849-5.188-0.183-0.156-9.744 4.177c0.052 0.077 0.099 0.161 0.14 0.249zM7.989 19.416l-0.167-0.165-3.38 1.389 1.64 1.631 1.865-2.896zM10.219 19.948c-0.235 0.109-0.495 0.161-0.749 0.167-0.209 0-0.407-0.032-0.605-0.099l-2.057 2.995 9.021 8.937 1.588-1.579zM8.953 16.427c0.135-0.015 0.276-0.015 0.416 0 0.152-0.020 0.308-0.020 0.459 0l3.645-5.667c-0.285-0.296-0.443-0.692-0.443-1.099-0.015-0.12-0.015-0.245 0-0.364l-4.427-2.047-3.563 3.468zM9.355 6.557l4.317 1.839c0.312-0.26 0.708-0.396 1.115-0.38 0.177 0.005 0.353 0.031 0.525 0.083l3.371-5.26-2.787-2.787-6.541 6.421c0 0-0.011 0.068 0 0.084zM7.74 17.443c0.088-0.161 0.197-0.312 0.323-0.448l-3.693-5.511-4.369 4.245zM7.391 18.375l-6.683-1.396 2.937 2.98 3.704-1.573z"
                  />
                </svg>
                 Netlify
              </button>
            </div>
          </div>
        </div>
        <p
          class="px-8 pb-4 text-center text-sm text-muted-foreground"
        >
          <a
            class="hover:text-brand underline underline-offset-4"
            href="/register"
          >
            Don't have an account? Sign Up
          </a>
        </p>
      </div>
    </div>
    <div
      aria-label="Notifications (F8)"
      role="region"
      style="pointer-events: none;"
      tabindex="-1"
    >
      <ol
        class="fixed top-0 z-[100] flex max-h-screen w-full flex-col-reverse p-4 sm:bottom-0 sm:right-0 sm:top-auto sm:flex-col md:max-w-[420px]"
        tabindex="-1"
      />
    </div>
  </div>
</DocumentFragment>
`;
